server/mediaserver.go

```diff
@@ -481,12 +287,88 @@ func endRTMPStreamHandler(s *LivepeerServer) func(url *url.URL, rtmpStrm stream.
 		if monitor.Enabled {
 			monitor.LogStreamEndedEvent(cxn.nonce)
 		}
+		cxn.eof <- struct{}{}
 		delete(s.rtmpConnections, mid)
 
 		return nil
 	}
 }
 
+func (s *LivepeerServer) startSession(cxn *rtmpConnection) *BroadcastSession {
+
+	mid := rtmpManifestID(cxn.stream)
+	cpl := cxn.pl
+	var sess *BroadcastSession
+
+	// Connect to the orchestrator. If it fails, retry for as long
+	// as the RTMP stream is alive
+	broadcastFunc := func() error {
+		var err error
+		sess, err = selectOrchestrator(s.LivepeerNode, cpl)
+		if err == ErrDiscovery {
+			return nil // discovery disabled, don't retry
+		} else if err != nil {
+			// Should be logged upstream
+		}
+		s.connectionLock.RLock()
+		defer s.connectionLock.RUnlock()
+		_, active := s.rtmpConnections[mid]
+		if active {
+			return err
+		}
+		return nil // stop if inactive
+	}
+	expb := backoff.NewExponentialBackOff()
+	expb.MaxInterval = BroadcastRetry
+	expb.MaxElapsedTime = 0
+	backoff.Retry(broadcastFunc, expb)
+	return sess
+}
+
+func (s *LivepeerServer) startSessionListener(cxn *rtmpConnection) {
+	mut := cxn.lock
+	sem := make(chan struct{}, 1)
+	runStartSession := func() {
+		mut.Lock()
```

Comment:
Yes, because the following `s.startSession` could take some time (it ends up invoking GetOrchestrator RPC, potentially multiple times). We don't want to block the `cxn` in other places that acquire the lock, such as during `processSegment` which acquires the lock for each segment, so every 2 seconds.

Added some comments clarifying this area of the code, including the connectionLock